SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestHandoverReplication/SpatialTestHandoverActorComponentReplication.cpp

```diff
@@ -0,0 +1,195 @@
+}
+
+void ASpatialTestHandoverActorComponentReplication::GetLifetimeReplicatedProps(TArray<FLifetimeProperty>& OutLifetimeProps) const
+{
+	Super::GetLifetimeReplicatedProps(OutLifetimeProps);
+
+	DOREPLIFETIME(ThisClass, HandoverCube);
+}
+
+void ASpatialTestHandoverActorComponentReplication::PrepareTest()
+{
+	Super::PrepareTest();
+
+	AddStep(TEXT("Check initial settings"), FWorkerDefinition::AllServers, nullptr, [this]() {
+		ASpatialWorldSettings* WorldSettings = Cast<ASpatialWorldSettings>(GetWorld()->GetWorldSettings());
+
+		if (AssertIsValid(WorldSettings, TEXT("World Settings of correct type")))
+		{
+			AssertTrue(IsValid(WorldSettings->GetMultiWorkerSettingsClass())
+						   && WorldSettings->GetMultiWorkerSettingsClass()->IsChildOf<USpatialTestHandoverReplicationMultiWorkerSettings>(),
+					   TEXT("MultiWorkerSettings should be of class USpatialTestHandoverReplicationMultiWorkerSettings"));
+		}
+
+		FinishStep();
+	});
+
+	AddStep(TEXT("Server 1 spawns a HandoverCube"), FWorkerDefinition::Server(1), nullptr, [this]() {
+		HandoverCube = GetWorld()->SpawnActor<AHandoverReplicationTestCube>(HandoverReplicationTestValues::Server1Position,
+																			FRotator::ZeroRotator, FActorSpawnParameters());
+		SaveHandoverCube(HandoverCube);
+		RegisterAutoDestroyActor(HandoverCube);
+		FinishStep();
+	});
+
+	auto AddWaitingStep = [this](const FString& StepName, const FWorkerDefinition& WorkerDefinition, TFunction<void()> TickFunction) {
+		constexpr float StepTimeLimit = 10.0f;
+
+		AddStep(
+			StepName, WorkerDefinition, nullptr, nullptr,
+			[this, TickFunction](float) {
+				TickFunction();
+				FinishStep();
+			},
+			StepTimeLimit);
+	};
+
+	AddWaitingStep(TEXT("Wait until the cube is synced with all servers"), FWorkerDefinition::AllServers, [this]() {
+		RequireTrue(IsValid(HandoverCube), TEXT("Server received the cube"));
+	});
+
+	AddStep(TEXT("Modify values on the Cube to non-default values"), FWorkerDefinition::Server(1), nullptr, [this]() {
+		HandoverCube->SetTestValues(HandoverReplicationTestValues::UpdatedTestPropertyValue);
+		HandoverCube->TestStage = EHandoverReplicationTestStage::ChangeValuesToDefaultOnGainingAuthority;
+		FinishStep();
+	});
+
+	AddWaitingStep(TEXT("Wait until updated values are received on all servers"), FWorkerDefinition::AllServers, [this]() {
+		HandoverCube->RequireTestValues(this, HandoverReplicationTestValues::UpdatedTestPropertyValue,
+										TEXT("Non-default value received on the server"));
+	});
+
+	// This step will trigger value change to the default one, after which the cube would be transported back to Server 1's authority area
+	AddStep(TEXT("Move Cube to Server 2's authority area"), FWorkerDefinition::Server(1), nullptr, [this]() {
+		HandoverCube->SetActorLocation(HandoverReplicationTestValues::Server2Position);
+		FinishStep();
+	});
+
+	AddWaitingStep(TEXT("Wait until value is reverted to default on all servers"), FWorkerDefinition::AllServers, [this]() {
```

Comment:
I had this as an initial version but the issue doesn't reliably reproduce. It's due to the shadow data being filled on the first call to `USpatialActorChannel::ReplicateActor`, which may or may not be called in certain conditions - for instance, if we rate limit a server, it might call `ReplicateActor` for only a subset of actors.

I can either add a comment detailing hows and whys, or add some more indirection to improve code readability, say:

```
class AHandoverReplicationTestCube
{
	DECLARE_DELEGATE(FOnAuthorityGained);
	FOnAuthorityGained OnAuthorityGainedDelegate;
};

void AHandoverReplicationTestCube::OnAuthorityGained()
{
	OnAuthorityGainedDelegate.Execute();
}

void PrepareTest()
{
	//...

	AddStep(TEXT("Modify HandoverCube's values once Server 2 gains authority over it"), FWorkerDefinition::Server(2), [](){ HandoverCube->OnAuthorityGainedDelegate.BindLambda([](){HandoverCube->SetValues(ModifiedValue);})});
}
```

This makes for significantly more readable code than what I have right now; as a contra - adds more scaffolding that's not 100% required.

Reply:
Fair enough, yeah maybe just a comment in between which directs the reader to the next steps in the test!